File: examples/chapter08/e05_kl_divergence/rust/src/main.rs
use burn::backend::NdArray;
use burn::nn::loss::{KLDivLossConfig, Reduction};
use burn::tensor::activation::{log_softmax, softmax};
use burn::tensor::{Tensor, TensorData};

type B = NdArray;

fn main() {
    let device = Default::default();

    let student_logits = Tensor::<B, 2>::from_data(
        TensorData::from([
            [2.0, 0.5, -1.0, 0.0],
            [0.1, 1.2, 0.3, -0.5],
            [-0.2, 0.4, 1.5, 0.8],
        ]),
        &device,
    );

    let teacher_logits = Tensor::<B, 2>::from_data(
        TensorData::from([
            [2.5, 0.2, -0.8, 0.1],
            [-0.1, 1.6, 0.2, -0.3],
            [0.0, 0.7, 1.2, 0.4],
        ]),
        &device,
    );

    // KL divergence expects log-probabilities for predictions.
    let student = log_softmax(student_logits, 1); // KLDiv wants LOG-probabilities
    let teacher = softmax(teacher_logits, 1); // the target is a distribution

    let loss = KLDivLossConfig::new()
        .init()
        .forward(student, teacher, Reduction::Auto); // Auto == batchmean

    println!("loss = {}", loss.into_scalar()); // ~0.034
}

File: examples/chapter08/e05_kl_divergence/python/main.py
import torch
import torch.nn.functional as F

# Teacher and student scores for three samples across four classes.
student_logits = torch.tensor(
    [
        [2.0, 0.5, -1.0, 0.0],
        [0.1, 1.2, 0.3, -0.5],
        [-0.2, 0.4, 1.5, 0.8],
    ],
)

teacher_logits = torch.tensor(
    [
        [2.5, 0.2, -0.8, 0.1],
        [-0.1, 1.6, 0.2, -0.3],
        [0.0, 0.7, 1.2, 0.4],
    ],
)

# KLDiv expects log-probabilities for input and probabilities for target.
student = F.log_softmax(student_logits, dim=1)     # log-probs
teacher = F.softmax(teacher_logits, dim=1)         # probs

loss = F.kl_div(student, teacher, reduction="batchmean")

print(f"loss = {loss.item()}")  # ~0.034
